views-react/src/main/java/io/micronaut/views/react/util/BeanPool.java

```diff
@@ -0,0 +1,180 @@
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.views.react.util;
+
+import io.micronaut.core.annotation.Internal;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.lang.ref.SoftReference;
+import java.util.LinkedList;
+import java.util.function.Supplier;
+
+/**
+ * <p>An object pool for beans that are both expensive to construct and not thread safe.</p>
+ *
+ * <p>
+ * The pool vends them temporarily to threads that need them when they call {@link #checkOut()}.
+ * This pool is designed to work well with virtual threads, by avoiding thread locals or any
+ * assumptions about how many distinct threads will request objects. If all the beans are currently
+ * checked out and a new bean is requested it will be created using the factory provided to the
+ * constructor. Therefore the number of beans created and available in the pool should grow to match
+ * the general rate of contention, rather than how many threads are in use.
+ * </p>
+ *
+ * <p>
+ * Beans in the pool can be atomically cleared and closed (if they implement {@link Closeable}).
+ * Any beans checked out when {@link #clear()} is called will remain untouched, but when they are
+ * checked back in they will be closed and discarded at that point.
+ * </p>
+ *
+ * <p>
+ * <b>Note:</b> this pool doesn't release objects in the background. If you have
+ * a sudden spike of traffic that drives many checkouts, memory usage may grow significantly
+ * and not be released. Fixing this would be a good future improvement to the pool.
+ * </p>
+ *
+ * @param <T> The type of the bean that being pooled.
+ */
+@Internal
+public class BeanPool<T> {
+    // TODO: Use @Scheduled to occasionally clear out beans that weren't accessed for a while to recover from traffic spikes.
+
+    private final Supplier<T> factory;
+
+    // Synchronized on 'this'.
+    private final LinkedList<SoftReference<PoolEntry>> pool = new LinkedList<>();
+    private int versionCounter = 0;   // File reloads.
+
+    /**
+     * Constructs an object pool that uses the given factory to create new entries.
+     *
+     * @param factory Used to create new entries when the pool is empty. Must be thread safe.
+     */
+    public BeanPool(Supplier<T> factory) {
+        this.factory = factory;
+    }
+
+    /**
+     * Returns a cached bean or creates a new one. Call {@link Handle#get()} to obtain the
+     * underlying object.
+     *
+     * @return a wrapper for the entry that you should pass to {@link #checkIn(Handle)} when you're
+     * done with it to put it back into the pool.
+     */
+    public synchronized Handle<T> checkOut() {
```

Comment:
I would prefer to use a lambda function to prevent accident missed close:

```
beanPool.<ResultType>useContext(ctx -> {
// use context
return result;
});
```

Reply:
Done